### ISSUE TITLE
SSHClientConnection.run closes stderr

### ISSUE THREAD

**Issue by stuz5000** [ronf/asyncssh#521]
It looks like SSHClientConnection.run closes stderr.

```

        async with asyncssh.connect(**kwargs) as conn:
            print("11==-- closed=", sys.stderr.closed)    # closed=False
            result = await conn.run(command=shellCommand,
                                    check=True,
                                    stdout=stdout,
                                    stderr=stderr,
                                    stdin=StringIO(stdinText)
                                    )
            print("11==-- closed=", sys.stderr.closed)    # closed=True
    
```

I assume:
 - This is a common use case
 - I shouldn't have to wrap stderr and mock close to avoid close being called.

Is there some way to call run that avoids .close being called? Or is this a bug?

**Comment by ronf** [comment]
This is by design - the `run()` method is intended to be used when you don't need to interactively perform I/O on the session. You can provide static input and/or redirect stdin/stdout/stderr, but the session will be ended before `run()` returns, providing the resulting output and exit status, or raising an exception if you get a non-zero exit status and set `check=True`.

If you want to do interactive I/O, take a look at `create_process()` instead of `run()`. That gives you complete control over reading and writing input and output, and deciding when to send EOF or close the connection. You can also refer to the example at https://asyncssh.readthedocs.io/en/latest/#interactive-input.

**Comment by stuz5000** [comment]
Hmm....  create_process also seems to close stderr:

```
        async with asyncssh.connect(**kwargs) as conn:
            print("11==-- closed=", sys.stderr.closed)
            stdout = StringIO()
            async with conn.create_process(command=shellCommand,
                                     stdout=stdout,
                                     stderr=sys.stderr,
                                     stdin=StringIO(stdinText)
                                     ) as process:
                result = await process.wait(timeout=timeout)
                print("11==-- closed=", sys.stderr.closed)                # True
                return stdout.getvalue()
```

**Comment by stuz5000** [comment]
Hmm....  create_process also seems to close stderr:

```
        async with asyncssh.connect(**kwargs) as conn:
            print("11==-- closed=", sys.stderr.closed)
            stdout = StringIO()
            async with conn.create_process(command=shellCommand,
                                     stdout=stdout,
                                     stderr=sys.stderr,
                                     stdin=StringIO(stdinText)
                                     ) as process:
                result = await process.wait(timeout=timeout)
                print("11==-- closed=", sys.stderr.closed)                # True
                return stdout.getvalue()
```

**Comment by stuz5000** [comment]
Hmm....  create_process also seems to close stderr:

```
        async with asyncssh.connect(**kwargs) as conn:
            print("11==-- closed=", sys.stderr.closed)
            stdout = StringIO()
            async with conn.create_process(command=shellCommand,
                                     stdout=stdout,
                                     stderr=sys.stderr,
                                     stdin=StringIO(stdinText)
                                     ) as process:
                result = await process.wait(timeout=timeout)
                print("11==-- closed=", sys.stderr.closed)                # True
                return stdout.getvalue()
```

**Comment by stuz5000** [comment]
OK. I think I figured it out:
```
        stdout = StringIO()
        stderr = StringIO()

        def collect(outs):
            stdout_data, stderr_data = outs
            if stderr_data:
                sys.stderr.write(stderr_data)
                stderr.write(stderr_data)
            if stdout_data:
                stdout.write(stdout_data)

        async with asyncssh.connect(**kwargs) as conn:
            async with conn.create_process(command=shellCommand) as process:
                process.stdin.write(stdinText)
                process.stdin.write_eof()
                collect(await process.communicate())
                await process.wait(check=True)
                collect(process.collect_output())

                if check and process.exit_status:
                    raise asyncssh.ProcessError(process.env, process.command, process.subsystem,
                                                process.exit_status, process.exit_signal,
                                                process.returncode, stdout.getvalue(), stderr.getvalue())
                else:
                    return asyncssh.SSHCompletedProcess(process.env, process.command, process.subsystem,
                                                        process.exit_status, process.exit_signal,
                                                        process.returncode,
                                                        stdout.getvalue(),
                                                        stderr.getvalue())
```

I tried sending lines incrementally to better interleave stdout/stderr and to be able to see which step of the input failed.
However, this leads `.communicate`  to hang. (On blank lines?)

**Comment by ronf** [comment]
If you use `communicate()`, you are going to run into the same problem as you did with `run()`. As documented, it waits for the process to finish before returning the tuple of output from stdout and output from stderr. Also, I see you are calling `wait()`, which is also going to block even if you weren't using `communicate()`.  If you want to read stdout and stderr in parallel, you can start two different asyncio tasks and either let them run independently or have each task do something like a readline() to read and return just a single line each time, waiting for one of the two to complete and then issuing a new readline() after processing the output and waiting on both tasks again. Alternately, if you don't care about distinguishing where the output came from, you can tell AsyncSSH to combine stdout and stderr all into stdout so there's only one stream to read. In fact, when allocating a PTY, there will only ever be output to stdout and stderr will always be empty.

Perhaps it would be good to back up and ask more specifically what you are trying to do.

**Comment by ronf** [comment]
Also, I just noticed something else -- is your concern here specifically about closing sys.stderr and not stderr on the SSH channel? If so, that is happening because you are redirecting output to sys.stderr and AsyncSSH passes through the EOF from the remote process. If you collect up the output yourself and write it to stderr, it shouldn't be closed.

Another option would be to duplicate the file descriptor for stdout/stderr and pass those new file descriptors into the redirect code. That wya, when the file descriptor is closed, it won't close the original sys.stdout or sys.stderr.

There's also a `send_eof` argument when setting up redirects that can also be used to suppress passing through EOF, but right now I think that only works when the target is an SSH channel. Code changes would be required to allow `send_eof` to work on other redirect targets like files. This seems like it might be a good change, but I'd need to look more closely at the code to see how complex a change that would be, as there are many different forms of output targets supported right now.

**Comment by ronf** [comment]
Assume you don't need to provide interactive input and the only issue is closing of sys.stderr, you should be able to go back to using `run()` here, with the only change being not to redirect stdout/stderr in the `run()` call. Instead, when it completes, do the checks to see if the returned SSHCompletedProcess has any stdout or stderr data collected, and write to the appropriate place.

The only downside to this is that you need to make sure the output doesn't get too large, as that can cause the output to block and the `run()` call to hang. This will be an issue with the solution you proposed here as well, though, or anything which doesn't incrementally read the output as it is being generated.

**Comment by ronf** [comment]
I haven't specifically tested it, but it looks to me like you could simplify what you have here down to:

```python
        async with asyncssh.connect(**kwargs) as conn:
            return conn.run(input=stdinText, command=shellCommand)
```

As long as you don't pass in sys.stdout or sys.stderr as stdout or stderr, the call to run() shouldn't close either of these. If you want the output to be sent to sys.stdout and sys.stderr without closing them, you can just write that output yourself using the values in the returned SSHCompletedProcess. In other words, something like:

```python
        async with asyncssh.connect(**kwargs) as conn:
            result = conn.run(input=stdinText, command=shellCommand)
            sys.stdout.write(result.stdout)
            sys.stderr.write(result.stderr)
```

If both are sent to the terminal, the output won't be interleaved, but if you want that, you could do something like:

```python
        async with asyncssh.connect(**kwargs) as conn:
            result = conn.run(input=stdinText, command=shellCommand, stderr=asyncssh.STDOUT)
            sys.stdout.write(result.stdout)
```
By

**Comment by stuz5000** [comment]
Thanks. I do sometimes want interactive input though. And doesn't stderr=asyncssh.STDOUT smush remote's stderr in to the local process's stdout?

An ideal interface for me would allow me to write:
```
  conn.run(onStderr=sys.stdin.write, onStderr=sys.stderr.write)
```
and allow me to either capture or interactively print output without having to drive the loop.  

Currently, I'm using the following for interactive sessions:
```
        async with asyncssh.connect(**kwargs) as conn:
            return await conn.run(
                command=shellCommand,
                term_type='xterm', stdin=sys.stdin,
                stdout=sys.stdout, stderr=sys.stderr,          
            )
```
but it has the side effect of closing stderr. Yikes!

**Comment by ronf** [comment]
If you prefer a callback-based API, AsyncSSH provides that. In fact, all of the streaming calls are implemented based on callbacks. You can find an example of this at https://asyncssh.readthedocs.io/en/latest/#callback-example. You can create your own SSHClientSession subclass, which you pass in as an argument to `create_session()`. You can optionally subclass SSHClient if you like as well, to implement callbacks at the connection level.

Regarding use of PTYs and separating stdout and stderr, it's not strictly speaking a limitation of the SSH protocol, but the nature of PTYs and TTYs is that you only get one input channel and one output channel on a TTY. So, pretty much all SSH server implementations using real PTYs/TTYs will end up combining output of stdout and stderr once a PTY is involved (which in the case of AsyncSSH can be triggered either explicitly or indirectly by setting `term_type`.

The reason stderr is closed in your bottom example is that you are still passing in sys.stderr as the stderr argument. For now, if you use redirects like this, the thing you are redirecting to will be sent an EOF when an EOF is received on the SSH session. If you leave stdout/stderr redirected to a PIPE (the default) and copy the data from that to sys.stderr yourself, sys.stderr won't be closed.

**Comment by stuz5000** [comment]
Thank you.
But this route doesn't seem to support interactive stdin.  (At least, the docs and example seem silent on it):
https://asyncssh.readthedocs.io/en/latest/api.html#asyncssh.SSHClientSession

Is it possible to start an interactive shell and receive the remote pty output and not have the any of stdin/stderr/stdout closed at the end of the session?

I only see conn.run taking stdin as a parameter:
```
        async with asyncssh.connect(**kwargs) as conn:
            return await conn.run(
                command=shellCommand,
                term_type='xterm',
                stdin=sys.stdin
            )
```

But the example https://asyncssh.readthedocs.io/en/latest/#callback-example
it's not obvious how to provide the overrides:
```
class MySSHClientSession(asyncssh.SSHClientSession):

class MySSHClient(asyncssh.SSHClient)
```
to prevent stderr being closed.

Neither SSHClientSession and SSHClient provide create_process or appear to deal with stdin.

**Comment by ronf** [comment]
There's no automatic redirect support in the callback-based APIs, but you have access to write() to stdin, and you will get callbacks every time data arrives on stdout or stderr, at which point you can do what you want with it (writing it to sys.stdout/sys.stderr, for instance, or capturing it somewhere) without having to do explicit reads.

For instance, consider the following:

```python
import asyncio, asyncssh, sys

class MySSHClientSession(asyncssh.SSHClientSession):
    def connection_made(self, chan: asyncssh.SSHClientChannel) -> None:
        self._chan = chan

    def session_started(self) -> None:
        self._chan.write('ls abc\n')
        self._chan.write_eof()

    def data_received(self, data: str, datatype: asyncssh.DataType) -> None:
        if datatype == asyncssh.EXTENDED_DATA_STDERR:
            sys.stderr.write(data)
        else:
            sys.stdout.write(data)

async def run_client() -> None:
    async with asyncssh.connect('localhost') as conn:
        chan, _ = await conn.create_session(MySSHClientSession)
        await chan.wait_closed()

try:
    asyncio.run(run_client())
except (OSError, asyncssh.Error) as exc:
    sys.exit('SSH connection failed: ' + str(exc))
```

This opens an interactive shell and writes "ls abc\n" as a command to it when the session opens, when the session_started() callback is called. As output comes out, the data_received() callback is called on each block of data, with an argument indicating if that output is associated with stdout (datatype of `None`) or stderr (datatype of `asyncssh.EXTENDED_DATA_STDERR`), copying that output to either sys.stdout or sys.stderr. 

To access the channel (such as when wanting to write to stdin), you need to implement the connection_made() callback and have it keep a copy of the `chan` argument passed to it. You can see all the available methods for that in [SSHClientChannel](https://asyncssh.readthedocs.io/en/latest/api.html#sshclientchannel). You can see all the possible callbacks you can implement in [SSHClientSession](https://asyncssh.readthedocs.io/en/latest/api.html#sshclientsession).

You can pass arguments like term_type to conn.create_session(), but be warned that not only does that collapse stdout and stderr but it also adds things like echoing your input and outputting shell prompts. As a result, it makes it much harder to cleanly read the output from any commands you might run. Also, you'll need to do something like writing 'exit\n' to exit the shell, rather than calling write_eof(). Unless you really need a TTY to run whatever commands you are trying to run, you're probably better off without that.

**Comment by stuz5000** [comment]
Many thanks. That was very helpful.
I managed to resolve the stdin echo issue by setting termios.TIOCSCTTY.
Shell history/edit control with the arrow keys and CTRL-D also seems to work as expected now also (e.g. can CTRL-D to exit `cat` but maintain the connection).

Sweet!

```
        async def stdinForwarder(remoteStdin: asyncssh.SSHClientChannel):
            loop = asyncio.get_event_loop()
            stdinAsyncReaders = asyncio.StreamReader()
            protocol = asyncio.StreamReaderProtocol(stdinAsyncReaders)
            await loop.connect_read_pipe(lambda: protocol, sys.stdin)
            fd = sys.stdin.fileno()
            try:
                old = termios.tcgetattr(fd)
                new = termios.tcgetattr(fd)
                new[3] = new[3] & ~termios.TIOCSCTTY

                while True:
                    termios.tcsetattr(fd, termios.TCSADRAIN, new)
                    res = await stdinAsyncReaders.read(10)
                    if not res:
                        break
                    remoteStdin.write(res.decode('utf8'))
            except KeyboardInterrupt:
                pass
            except Exception:
                debug.reportException()
            finally:
                termios.tcsetattr(fd, termios.TCSADRAIN, old)
                if not remoteStdin.is_closing():
                    if remoteStdin.can_write_eof():
                        remoteStdin.write_eof()
                    remoteStdin.close()

        class MySSHClientSession(asyncssh.SSHClientSession):
            def connection_made(self, chan: asyncssh.SSHClientChannel) -> None:
                self._remoteStdin = chan
                return super().connection_made(chan)

            def session_started(self) -> None:
                # self._remoteStdin.write('ls abc\n')
                # self._remoteStdin.write_eof()
                # return

                async def f():
                    self._stdInWriterTask = asyncio.create_task(stdinForwarder(self._remoteStdin))

                asyncio.run_coroutine_threadsafe(f(), asyncio.get_event_loop())
                return super().session_started()

            def data_received(self, data: str, datatype: asyncssh.DataType) -> None:
                if datatype == asyncssh.EXTENDED_DATA_STDERR:
                    print(data, end='', file=sys.stderr)
                    sys.stderr.flush()
                else:
                    print(data, end='')
                    sys.stdout.flush()

            def connection_lost(self, exc: Optional[Exception]) -> None:
                task = None
                try:
                    task, self._stdInWriterTask = self._stdInWriterTask, None
                except:
                    pass

                if task is not None:
                    task.cancel()

                if exc:
                    print('SSH session error: ' + str(exc), file=sys.stderr)

        class MySSHClient(asyncssh.SSHClient):
            def connection_made(self, conn: asyncssh.SSHClientConnection) -> None:
                print('Connection made to %s.' % conn.get_extra_info('peername')[0])

            def auth_completed(self) -> None:
                print('Authentication successful.')


        conn, client = await asyncssh.create_connection(MySSHClient,
                                                        **kwargs)
        async with conn:
            chan, session = await conn.create_session(MySSHClientSession,
                                                      command=shellCommand,
                                                      term_type='xterm',
                                                      # term_type='auto',
                                                      )
            await chan.wait_closed()
```

**Comment by ronf** [comment]
Happy to help! Glad you got it working...

If you are trying to use the real stdin and it's a TTY, I can see how what you're doing here would work for disabling local echo. However, once you request a PTY on the SSH session, I'd expect the remote TTY to do echoing as well, and that would be harder to shut off without running something like the "stty" command in the remote shell. Also, once the remote shell sees that you have a TTY, it will begin doing things like printing prompts, so you may need to disable those or filter them out of your output.

I'm experimenting with some form of "keep open" option for output redirection, so that you could potentially do something like this with `run()` and its redirect feature. There are some edge cases I still need to think about, as well as the best way to name this vs. the existing "send_eof" argument, but I should have something in a few days if you wanted to try it out. I'll let you know here when I do.

**Comment by stuz5000** [comment]
Hmmm... in the above

Let's say I just want to send:

```
#bash -l running remotely via MySSHClientSession. Sending:

                    remoteStdin.write("ls -l\n")
                    if remoteStdin.can_write_eof():
                        remoteStdin.write_eof()
                    //remoteStdin.close()
```
My understanding is that sending eof should cause the remote shell to disconnect, but connection_lost is not called.
If I call remoteStdin.close however, connection_lost is called before the "ls -l" output response is received (at least, it doesn't make it to the data_received function).

In the create_process examples, we could write:
```
        async with asyncssh.connect(**kwargs) as conn:
            async with conn.create_process(command=shellCommand) as process:
                process.stdin.write(stdinText)   # ls -l
                process.stdin.write_eof()
                collectOutput(await process.communicate())
                
                await process.wait(check=True)
                collectOutput(process.collect_output())    
```

Is it possible to do the same via the SSHClient/SSHClientSession route? (i.e. Wait for execution to complete collect the received remote output?)

**Comment by ronf** [comment]
If you do the write() and write_eof() as you show here without calling close(), you should be able to collect the output in your data_received() callback, and you'll know that it is done when you get back an eof_received() callback. At that point, you can call close() on the session.

I would expect the write() calls to be on the SSH channel object returned by create_session(), though. There isn't anything like a stdin/stdout/stderr object in this case. In the case where you are creating an SSH client session, write() with no "datatype" argument writes to stdin, and there are no other datatypes in that direction. However, if you were using this kind of channel object on the server side, write() to that channel would be output to stdout and write() with the datatype of asyncssh.EXTENDED_DATA_STDERR would be output to stderr.

Your data_received() callback on the client side passes in a datatype argument to you which follows the same pattern for distinguishing stdout vs. stderr.

As for connection_lost(), I don't think that'll ever be called when you initiate the close. However, if the remote end closes first, it will be called, probably after a call to your eof_received() callback. If it's a clean close, the exc argument will be None in that case.

The remote shell should exit when you do the write_eof(), but you could also potentially write something like "exit\n" explicitly to tell it to exit, and that should trigger the same eof_received() callback after all the output is written (and eventually a call to connection_lost()).

**Comment by ronf** [comment]
Well, it took a bit longer than I expected, but I was finally able to check in a change which allows you to control whether redirected output targets are closed when EOF is received on an SSH channel. When this new "recv_eof" argument is set to `false`, you can reuse a redirect target to run several commands and even intermix your own output in between redirects. This is meant to parallel the current "send_eof" argument, which decides whether an EOF received from a redirected input source will be forwarded to the SSH channel or not, allowing multiple inputs to be sequentially redirected to a channel (along with other dynamically generated input).

This new "recv_eof" option is available on the `create_process()` call to start a new process and also on the `redirect()` function of a created process, so that this choice can be made each time a redirect is updated. The changes are available in the "develop" branch as commit b391005.

For this application, you might still be better to stick to callbacks if you want to handle stdout and stderr in parallel without creating your own independent reading threads. However, in cases where using processes makes sense, this new option should make sending output from multiple commands to the same redirect target much easier.

**Comment by stuz5000** [comment]
Many thanks!

**Comment by ronf** [comment]
This new "recv_eof" option is now available in AsyncSSH 2.13.0.
